karabo_data/geometry2.py

```diff
@@ -0,0 +1,326 @@
+                tiles.append(AGIPDGeometryFragment(
+                    corner_pos=np.array([corner_x, corner_y, 0.]),
+                    ss_vec=np.array([x_orient, 0, 0]),
+                    fs_vec=np.array([0, y_orient, 0]),
+                ))
+        return cls(modules)
+
+    @classmethod
+    def from_crystfel_geom(cls, filename):
+        geom_dict = load_crystfel_geometry(filename)
+        modules = []
+        for p in range(16):
+            tiles = []
+            modules.append(tiles)
+            for a in range(8):
+                d = geom_dict['panels']['p{}a{}'.format(p, a)]
+                tiles.append(AGIPDGeometryFragment.from_panel_dict(d))
+        return cls(modules)
+
+    def write_crystfel_geom(self, filename):
+        from . import __version__
+
+        panel_chunks = []
+        for p, module in enumerate(self.modules):
+            for a, fragment in enumerate(module):
+                panel_chunks.append(fragment.to_crystfel_geom(p, a))
+
+        with open(filename, 'w') as f:
+            f.write(CRYSTFEL_HEADER_TEMPLATE.format(version=__version__))
+            for chunk in panel_chunks:
+                f.write(chunk)
+
+    def inspect(self):
+        """Plot the 2D layout of this detector geometry.
+
+        Returns a matplotlib Figure object.
+        """
+        from matplotlib.backends.backend_agg import FigureCanvasAgg
+        from matplotlib.collections import PatchCollection
+        from matplotlib.figure import Figure
+        from matplotlib.patches import Polygon
+
+        fig = Figure((10, 10))
+        FigureCanvasAgg(fig)
+        ax = fig.add_subplot(1, 1, 1)
+
+        rects = []
+        for p, module in enumerate(self.modules):
+            for a, fragment in enumerate(module):
+                corners = fragment.corners()[:, :2]  # Drop the Z dimension
+
+                rects.append(Polygon(corners))
+
+                if a in {0, 7}:
+                    ax.text(*fragment.centre()[:2], str(a),
+                            verticalalignment='center',
+                            horizontalalignment='center')
+                elif a == 4:
+                    ax.text(*fragment.centre()[:2], 'p{}'.format(p),
+                            verticalalignment='center',
+                            horizontalalignment='center')
+
+        pc = PatchCollection(rects, facecolor=(0.75, 1., 0.75), edgecolor=None)
+        ax.add_collection(pc)
+
+        ax.hlines(0, -100, +100, colors='0.75', linewidths=2)
+        ax.vlines(0, -100, +100, colors='0.75', linewidths=2)
+
+        ax.set_title('AGIPD-1M detector geometry')
+        return fig
+
+    def position_all_modules(self, data):
+        """Assemble data from this detector according to where the pixels are.
+
+        Parameters
+        ----------
+
+        data : ndarray
+          The last three dimensions should be channelno, pixel_y, pixel_x
+          (lengths 16, 256, 256).
```

Comment:
`(16, 512, 128)`
aren't pixels x and y the other way around?

Reply:
x and y are confusing in terms of data; I've now switched these docstrings to CrystFELs fs/ss terminology.